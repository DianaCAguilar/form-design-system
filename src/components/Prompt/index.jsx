import React from 'react';
import PropTypes from 'prop-types';
import Dialog from 'components/Dialog';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';

const Prompt = (props) => (
  <Dialog
    content={props.desc ? <div id="a11y-dialog-desc">{props.desc}</div> : null}
    footerContent={
      <Flex justify="end" wrap>
        {props.secondaryButton && <FlexItem shrink>{props.secondaryButton}</FlexItem>}
        {props.primaryButton && <FlexItem shrink>{props.primaryButton}</FlexItem>}
      </Flex>
    }
    role="alertdialog"
    title={props.title}
    isOpen={props.isOpen}
    onDismiss={props.onDismiss}
    disablePortal={props.disablePortal}
  />
);

Prompt.propTypes = {
  /** Controls text in header */
  title: PropTypes.string,
  /** Controls string displayed in content area */
  desc: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  /** Represents primary button (MUST be an FDS Button) */
  primaryButton: PropTypes.any.isRequired,
  /** Represents secondary button (MUST be an FDS Button) */
  secondaryButton: PropTypes.any.isRequired,
  /** Controls whether modal (and backdrop) are visible or not */
  isOpen: PropTypes.bool,
  /**
   * Callback that user can pass in, to be conditionally fired when
   * user attempts to close modal. When defined, the modal close button appears
   */
  onDismiss: PropTypes.func,
  /** Keeps the Prompt from being rendered in a portal and renders it locally instead. */
  disablePortal: PropTypes.bool,
};

export default Prompt;
