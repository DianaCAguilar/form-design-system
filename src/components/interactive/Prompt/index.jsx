import React from 'react';
import PropTypes from 'prop-types';
import Dialog from '../Dialog';
import Button from '../Button';
import Flex from '../../layout/Flex';
import FlexItem from '../../layout/FlexItem';

const Reusable = (props) => (
  <div className="">
    <Flex justify="end" wrap>
      {props.secondaryButton && (
        <FlexItem shrink>
          <Button {...props.secondaryButton} />
        </FlexItem>
      )}
      <FlexItem shrink>
        <Button {...props.primaryButton} />
      </FlexItem>
    </Flex>
  </div>
);

Reusable.propTypes = {
  primaryButton: PropTypes.any.isRequired,
  secondaryButton: PropTypes.any,
};

const Prompt = (props) => (
  <Dialog
    footerContent={props.buttonPosition === 'footer' && <Reusable {...props} />}
    content={
      <React.Fragment>
        <div>{props.content}</div>
        {props.buttonPosition === 'content' && (
          <div className="margin--top--double">
            <Reusable {...props} />
          </div>
        )}
      </React.Fragment>
    }
    role="alertdialog"
    canDismiss={false}
    title={props.title}
    isOpen={props.isOpen}
    onDismiss={props.onDismiss}
  />
);

Prompt.defaultProps = {
  buttonPosition: 'content',
};

Prompt.propTypes = {
  title: PropTypes.string,
  content: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  primaryButton: PropTypes.any.isRequired,
  secondaryButton: PropTypes.any,
  isOpen: PropTypes.bool,
  onDismiss: PropTypes.func,
  buttonPosition: PropTypes.oneOf(['content', 'footer']),
};

export default Prompt;
