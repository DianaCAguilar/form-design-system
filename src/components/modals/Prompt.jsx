import React from 'react';
import PropTypes from 'prop-types';
import Dialog from './Dialog';
import Button from '../interactive/Button';
import Flex from '../layout/Flex';
import FlexItem from '../layout/FlexItem';

const Prompt = (props) => (
  <Dialog
    content={
      <React.Fragment>
        <div id="a11y-dialog-desc">{props.desc}</div>
        <div className="margin--top--double">
          <div className="">
            <Flex justify="end" wrap>
              {props.secondaryButton && (
                <FlexItem shrink>
                  <Button {...props.secondaryButton} />
                </FlexItem>
              )}
              <FlexItem shrink>
                <Button {...props.primaryButton} />
              </FlexItem>
            </Flex>
          </div>
        </div>
      </React.Fragment>
    }
    role="alertdialog"
    canDismiss={false}
    title={props.title}
    isOpen={props.isOpen}
    onDismiss={props.onDismiss}
  />
);

Prompt.defaultProps = {
  buttonPosition: 'content',
};

Prompt.propTypes = {
  title: PropTypes.string,
  desc: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  primaryButton: PropTypes.any.isRequired,
  secondaryButton: PropTypes.any.isRequired,
  isOpen: PropTypes.bool,
  onDismiss: PropTypes.func,
  buttonPosition: PropTypes.oneOf(['content', 'footer']),
};

export default Prompt;
