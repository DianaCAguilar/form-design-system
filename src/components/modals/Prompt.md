# Prompt

A confirm modal that displays a simple message, with 2 buttons to either confirm the action, or deny it (does not show close icon, and
can't be dismissed by pressing ESC or clicking backdrop)