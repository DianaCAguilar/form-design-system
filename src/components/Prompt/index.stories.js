import React from 'react';
import { action } from '@storybook/addon-actions';
import ImportPath from 'util/storybook-docs/ImportPath';
import Button from 'components/Button';

import {
  Title,
  Subtitle,
  Description,
  ArgsTable,
  PRIMARY_STORY,
} from '@storybook/addon-docs/blocks';

import Prompt from '.';

export const Primary = () => (
  <Prompt
    isOpen={true}
    desc="This change will be permanent and cannot be undone."
    title="Are you sure you want to delete this collection?"
    primaryButton={<Button isDestructive label="Delete" />}
    secondaryButton={
      <Button theme="ghost" label="Keep Collection" onClick={action('close modal')} />
    }
    onDismiss={action('close modal')}
  />
);

const Layout = () => (
  <>
    <Title />
    <ImportPath />
    <Subtitle />
    <Description />
    <ArgsTable story={PRIMARY_STORY} />
  </>
);

export default {
  component: Prompt,
  title: 'components/Prompt',
  parameters: {
    docs: {
      fds: {
        related: ['Dialog'],
      },
      page: Layout,
      description: {
        component:
          'A confirm modal that displays a simple message, with 2 buttons to either confirm the action, or deny it. **Please click "Canvas" to view examples**.',
      },
    },
  },
};
