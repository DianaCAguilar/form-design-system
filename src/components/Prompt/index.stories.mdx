import React from 'react';
import { action } from '@storybook/addon-actions';
import ImportPath2 from 'util/storybook-docs/ImportPath2';
import Button from "components/Button";

import {
  Title,
  Subtitle,
  Description,
  ArgsTable,
  PRIMARY_STORY,
} from '@storybook/addon-docs/blocks';

import Prompt from ".";

export const Primary = () => (
  <Prompt
    isOpen={true}
    desc="This change will be permanent and cannot be undone."
    title="Are you sure you want to delete this collection?"
    primaryButton={<Button isDestructive label="Delete" />}
    secondaryButton={<Button theme="ghost" label="Keep Collection" onClick={action("close modal")} />}
    onDismiss={action('close modal')}
  />
)

const Layout = () => (
  <>
    <Title />
    <ImportPath2 />
    <Subtitle />
    <Description />
    <ArgsTable story={PRIMARY_STORY} />
  </>
);

export default {
  component: Prompt,
  title: 'components/Prompt',
  parameters: {
    docs: {
      page: Layout,
      description: {
        component:
          'A confirm modal that displays a simple message, with 2 buttons to either confirm the action, or deny it. **Please click "Canvas" to view examples**.',
      },
    },
  },
};

