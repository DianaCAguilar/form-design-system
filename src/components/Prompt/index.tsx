import React, { useEffect, useRef, useState } from 'react';
import cc from 'classcat';
import Flex from 'components/Flex';
import { DenyIcon } from 'icons/react';
import { IconButton } from 'components';
import FlexItem from 'components/FlexItem';
import { ButtonProps } from 'components/Button';
import * as AlertDialog from '@radix-ui/react-alert-dialog';
import rafSchd from 'raf-schd';

export const isElementOverflowing = ({
  current,
}: React.MutableRefObject<HTMLDivElement | null>): boolean => {
  // Checking for current first is safer just in case,
  // and will also fail gracefully in tests
  if (current) {
    return current.scrollHeight > current.clientHeight;
  }
  return false;
};

export interface PromptProps {
  /** Controls text in header */
  title?: string;
  /** Controls string displayed in content area */
  desc?: React.ReactNode | string;
  /** Represents primary button (MUST be an FDS Button) */
  primaryButton: React.ReactElement<ButtonProps>;
  /** Represents secondary button (MUST be an FDS Button) */
  secondaryButton: React.ReactElement<ButtonProps>;
  /** Controls whether modal (and backdrop) are visible or not */
  isOpen?: boolean;
  /**
   * Callback that user can pass in, to be conditionally fired when
   * user attempts to close modal. When defined, the modal close button appears
   */
  onDismiss?: () => void;
  /** Keeps the Prompt from being rendered in a portal and renders it locally instead. */
  disablePortal?: boolean;
}

const Prompt = ({
  title,
  desc,
  primaryButton,
  secondaryButton,
  isOpen,
  onDismiss,
}: PromptProps): JSX.Element => {
  const contentEl = useRef<HTMLDivElement>(null);
  const [isOverflowing, setIsOverflowing] = useState(false);
  const schdSetIsOverflowing = rafSchd(setIsOverflowing);

  const handleResize = () => {
    schdSetIsOverflowing(isElementOverflowing(contentEl));
  };

  useEffect(() => {
    handleResize(); // needs to fire once immediately on mount
    window.addEventListener('resize', handleResize);
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  return (
    <AlertDialog.Root open={isOpen} onOpenChange={onDismiss}>
      <AlertDialog.Overlay className="dialog-overlay"></AlertDialog.Overlay>
      <div className="dialog-zIndex dialog-wrapper">
        <AlertDialog.Content
          // onPointerDownOutside={(event) => event.preventDefault()}
          asChild
          className="ease-in-out"
        >
          <div
            className="dialog elevation--3 border--focus--noTransition"
            role={'alertdialog'}
            aria-labelledby={title && 'a11y-dialog-title'}
            aria-describedby="a11y-dialog-desc"
            tabIndex={-1}
            aria-modal="true"
            style={{
              maxWidth: '500px',
              maxHeight: '80vh',
            }}
            data-testid="dialog"
          >
            {(title || onDismiss) && (
              <>
                <div className="dialog-header">
                  <div className="bgColor--white padding--all border--bottom">
                    <div>
                      <div className="padding--right--xl type--head4">
                        {title ? <span id="a11y-dialog-title">{title}</span> : '\u00A0'}{' '}
                        {/* There always needs to be something (even a space) in the header for display reasons */}
                      </div>
                    </div>
                    {onDismiss && (
                      <AlertDialog.Trigger asChild>
                        <div className="dialog-icon">
                          <IconButton Icon={DenyIcon} aria-label="Close" label="Close" />
                        </div>
                      </AlertDialog.Trigger>
                    )}
                  </div>
                </div>
              </>
            )}
            {desc && (
              <div className="dialog-content" ref={contentEl}>
                <div className="padding--all bgColor--white">
                  {desc ? <div id="a11y-dialog-desc">{desc}</div> : null}
                </div>
              </div>
            )}
            {
              <div className="dialog-footer">
                <div
                  className={cc([
                    {
                      'border--top': isOverflowing,
                    },
                    'bgColor--white',
                    'padding--all',
                  ])}
                >
                  <Flex justify="end">
                    {secondaryButton && <FlexItem shrink>{secondaryButton}</FlexItem>}
                    {primaryButton && <FlexItem shrink>{primaryButton}</FlexItem>}
                  </Flex>
                </div>
              </div>
            }
          </div>
        </AlertDialog.Content>
      </div>
    </AlertDialog.Root>
  );
};

export default Prompt;
