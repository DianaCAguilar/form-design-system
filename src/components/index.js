import Avatar from 'components/Avatar';
import Badge from 'components/Badge';
import Button from 'components/Button';
import ButtonGroup from 'components/ButtonGroup';
import Checkbox from 'components/Checkbox';
import Chip from 'components/Chip';
import DateInput from 'components/DateInput';
import DecoratedInput from 'components/DecoratedInput';
import Dialog from 'components/Dialog';
import DropdownButton from 'components/DropdownButton';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import FloatingAction from 'components/FloatingAction';
import Hscroll from 'components/Hscroll';
import IconButton from 'components/IconButton';
import IconInput from 'components/IconInput';
import InputGroup from 'components/InputGroup';
import Menu from 'components/Menu';
import Popover from 'components/Popover';
import Prompt from 'components/Prompt';
import Radio from 'components/Radio';
import Section from 'components/Section';
import StackedButton from 'components/StackedButton';
import TextInput from 'components/TextInput';
import Toaster from 'components/Toaster';
import Tooltip from 'components/Tooltip';
import Updatable from 'components/Updatable';

export {
  Avatar,
  Badge,
  Button,
  ButtonGroup,
  Checkbox,
  Chip,
  DateInput,
  DecoratedInput,
  Dialog,
  DropdownButton,
  Flex,
  FlexItem,
  FloatingAction,
  Hscroll,
  IconButton,
  IconInput,
  InputGroup,
  Menu,
  Popover,
  Prompt,
  Radio,
  Section,
  StackedButton,
  TextInput,
  Toaster,
  Tooltip,
  Updatable,
};
