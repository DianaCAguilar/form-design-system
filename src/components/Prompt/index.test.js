import React from 'react';
import { render, screen } from '@testing-library/react';
import { Button } from 'components';
import Prompt from '.';

describe('Prompt component', () => {
  it('tests that prompt renders properly', () => {
    render(
      <Prompt
        isOpen
        primaryButton={<Button label="foo" />}
        secondaryButton={<Button label="bar" />}
      />
    );

    expect(screen.getByText('foo')).toBeTruthy();
    expect(screen.getByText('bar')).toBeTruthy();
  });
});
