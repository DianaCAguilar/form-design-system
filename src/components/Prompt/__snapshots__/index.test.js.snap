// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prompt component matches snapshot (default props) 1`] = `
<Dialog
  footerContent={
    <Flex
      justify="end"
      wrap={true}
    >
      <FlexItem
        shrink={true}
      >
        <Button
          label="bar"
        />
      </FlexItem>
      <FlexItem
        shrink={true}
      >
        <Button
          label="foo"
        />
      </FlexItem>
    </Flex>
  }
/>
`;

exports[`Prompt component matches snapshot (default props) 2`] = `
<Dialog
  footerContent={
    <Flex
      justify="end"
      wrap={true}
    >
      <FlexItem
        shrink={true}
      >
        <Button
          label="bar"
        />
      </FlexItem>
      <FlexItem
        shrink={true}
      >
        <Button
          label="foo"
        />
      </FlexItem>
    </Flex>
  }
  isOpen={true}
/>
`;

exports[`Prompt component matches snapshot (set all props) 1`] = `
<Dialog
  content="desc"
  footerContent={
    <Flex
      justify="end"
      wrap={true}
    >
      <FlexItem
        shrink={true}
      >
        <Button
          label="bar"
        />
      </FlexItem>
      <FlexItem
        shrink={true}
      >
        <Button
          label="foo"
        />
      </FlexItem>
    </Flex>
  }
  isOpen={true}
  title="title"
/>
`;
